packages/devtools_app/lib/src/timeline/timeline_model.dart

```diff
@@ -11,27 +15,162 @@ import '../utils.dart';
+
+  static const selectedFrameIdKey = 'selectedFrameId';
+
+  FutureOr<double> get displayRefreshRate async =>
+      _displayRefreshRate ??= await serviceManager.getDisplayRefreshRate();
+
+  double _displayRefreshRate;
+
+  /// All frames currently visible in the timeline.
+  List<TimelineFrame> frames = [];
+
+  TimelineFrame selectedFrame;
+
+  String get selectedFrameId => selectedFrame?.id;
+
+  @override
+  Map<String, dynamic> get json => {
+        selectedFrameIdKey: selectedFrame?.id,
+        displayRefreshRateKey: _displayRefreshRate ?? defaultRefreshRate,
+      }..addAll(super.json);
+
+  @override
+  int get displayDepth =>
+      selectedFrame.uiEventFlow.depth + selectedFrame.gpuEventFlow.depth;
+
+  @override
+  void clear() {
+    super.clear();
+    frames.clear();
+    selectedFrame = null;
+    _displayRefreshRate = null;
+  }
+}
+
+class FullTimelineData extends TimelineData {
+  FullTimelineData({
+    List<Map<String, dynamic>> traceEvents,
+    TimelineEvent selectedEvent,
+    CpuProfileData cpuProfileData,
+    List<TimelineEvent> timelineEvents,
+  })  : timelineEvents = timelineEvents ?? [],
+        super(
+          traceEvents: traceEvents ?? [],
+          selectedEvent: selectedEvent,
+          cpuProfileData: cpuProfileData,
+        );
+
+  static const uiKey = 'UI';
+
+  static const gpuKey = 'GPU';
+
+  static const unknownKey = 'Unknown';
+
+  final List<TimelineEvent> timelineEvents;
+
+  final SplayTreeMap<String, List<TimelineEvent>> eventBuckets =
+      SplayTreeMap(eventBucketComparator);
+
+  TimeRange time = TimeRange();
+
+  @override
+  int get displayDepth {
+    if (_displayDepth != null) return _displayDepth;
+
+    if (eventBuckets.isEmpty) {
+      initializeEventBuckets();
+    }
+
+    int depth = 0;
+    for (List<TimelineEvent> bucket in eventBuckets.values) {
+      int bucketDisplayDepth = 0;
+      for (TimelineEvent event in bucket) {
+        bucketDisplayDepth = max(bucketDisplayDepth, event.displayDepth);
+      }
+      depth += bucketDisplayDepth;
+    }
+    return depth;
+  }
+
+  int _displayDepth;
+
+  void initializeEventBuckets() {
+    for (TimelineEvent event in timelineEvents) {
+      if (event.isAsyncEvent) {
+        eventBuckets.putIfAbsent(event.name, () => [])..add(event);
+      } else if (event.isUiEvent) {
+        eventBuckets.putIfAbsent(uiKey, () => [])..add(event);
+      } else if (event.isGpuEvent) {
+        eventBuckets.putIfAbsent(gpuKey, () => [])..add(event);
+      } else {
+        eventBuckets.putIfAbsent(unknownKey, () => [])..add(event);
+      }
+    }
+  }
+
+  @override
+  void clear() {
+    super.clear();
+    timelineEvents.clear();
+    eventBuckets.clear();
+    time = TimeRange();
+    _displayDepth = null;
+  }
+
+  @visibleForTesting
+  static int eventBucketComparator(String a, String b) {
+    if (a == b) return 0;
+
+    // Order Unknown buckets last.
+    if (a == unknownKey && b != unknownKey) return 1;
```

Comment:
the
b!= unknownkey and
a != unknownKey cases are not needed.
a != b 
from the previous case.

Reply:
done.